### PR TITLE
IFTAQCAA-19 - Add Product model into the project

### PR DESCRIPTION
**Pull request by dkrutenko**
moved methods from another classes, added new methods

### CODE REVIEW

**Comment by Nikodych** [comment]
sync with master

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.List;
+
+import static com.codeborne.selenide.CollectionCondition.sizeGreaterThan;
+import static com.codeborne.selenide.Condition.visible;
+import static com.codeborne.selenide.Selenide.$$x;
+import static com.codeborne.selenide.Selenide.$x;
+import static java.time.Duration.ofSeconds;
+import static java.util.stream.Collectors.toList;
+
+public class ProductModel extends BasePage<ProductModel> {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:15]: how it's could be a model? the model should represent an object, but not a page object

also, I think all things should be in SearchResultPage

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @Step("ProductModel: Get old price from first product")
+    public String getOldPriceFromFirstItem() {
+        return $x("//ul[@class='catalog-grid ng-star-inserted']/li[1]//div[@class='goods-tile__price--old price--gray ng-star-inserted']")
+                .shouldBe(visible, ofSeconds(12))
+                .getText()
+                .trim();
+    }
+
+    @Step("ProductModel: Get amount of response from first product")
+    public String getAmoutOfResponseFromAnyItem(String itemNumber) {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:44]: getNumberOfReviewsFromItemBy

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        return $x("//li[" + itemNumber + "]//div[@class='goods-tile__stars ng-star-inserted']/*[name() = 'svg']")
+                .shouldBe(visible, ofSeconds(12))
+                .getText()
+                .trim();
+    }
+
+    @Step("ProductModel: Click on 'add to cart' button on first product")
+    public ProductModel clickAddToCartOnAnyItem(String itemNumber) {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:52]: addProductToCartBy

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Step("ProductModel: Get old price from first product")
+    public String getOldPriceFromFirstItem() {
+        return $x("//ul[@class='catalog-grid ng-star-inserted']/li[1]//div[@class='goods-tile__price--old price--gray ng-star-inserted']")
+                .shouldBe(visible, ofSeconds(12))
+                .getText()
+                .trim();
+    }
+
+    @Step("ProductModel: Get amount of response from first product")
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:43]: for get and logic methods steps not needed

Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:43]: where these methods are used? this one and under it

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java (hunk 1 above)
Comment (dkrutenko) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:15]: Probably, then we dont need a productmodel...

Comment (dkrutenko) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:15]: I tought that ProductModel means some product construction with some functions that attach to any random product

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java (hunk 4 above)
Comment (dkrutenko) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:43]: They have never been used. But, maybe, in future they will

File: src/main/java/com/softserveinc/ita/models/ProductModel.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.softserveinc.ita.models;
+
+import lombok.Builder;
+
+@Builder
+public class ProductModel {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/models/ProductModel.java:6]: just Product

File: src/main/java/com/softserveinc/ita/models/repos/ProductRepo.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package com.softserveinc.ita.models.repos;
+
+import com.softserveinc.ita.models.ProductModel;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static com.softserveinc.ita.utils.NumberUtil.parseIntPrice;
+
+public class ProductRepo {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/models/repos/ProductRepo.java:8]: how it's could be the repo if it's some page object?

File: src/main/java/com/softserveinc/ita/models/repos/ProductRepo.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    public static int getPriceBeforeDiscount() {
+        return parseIntPrice($x("//p[contains(@class, 'product-prices__small')]").getText());
+    }
+
+    public static String getTitle() {
+        return $x("//h1[@class='product__title']").getText();
+    }
+
+    public ProductModel getProductInfo() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/models/repos/ProductRepo.java:22]: as I understand, all these actions on the product page, use that page object, repo it's something like data provider

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
+package com.softserveinc.ita.pageobjects;
 
-import com.softserveinc.ita.models.ProductModel;
+import com.softserveinc.ita.models.Product;
 
 import static com.codeborne.selenide.Selenide.$x;
 import static com.softserveinc.ita.utils.NumberUtil.parseIntPrice;
 
-public class ProductRepo {
+public class ProductModel {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:8]: why this class is needed? if it's related to a ProductPage ?

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java (hunk 8 above)
Comment (dkrutenko) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:8]: where i have to move them?

Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductModel.java:8]: all these methods probably should be in ProductPage, as I understand all operations occur there

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 9]
```diff
@@ -63,15 +66,66 @@ private String getActiveTabText() {
...
+    }
+
+    public int getKitPrice(int index) {
+        var kitPrice = $x(format("//li[@class = 'simple-slider__item ng-star-inserted'][%s]//div[@class = 'kits-price__coast']", index))
+                .shouldBe(visible)
+                .getText();
+
+        return parseIntPrice(kitPrice);
+    }
+}
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:110]: I think it's an extra bracket

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 10]
```diff
@@ -63,15 +66,66 @@ private String getActiveTabText() {
...
+
+    public static String getProductTitle() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:112]: non static

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 11]
```diff
@@ -63,15 +66,66 @@ private String getActiveTabText() {
...
+        return $x("//h1[@class='product__title']").getText();
+    }
+
+    public static int getProductPrice() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:116]: same

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 12]
```diff
@@ -63,15 +66,66 @@ private String getActiveTabText() {
...
+        return parseIntPrice($x("//p[contains(@class, 'product-prices__big')]").getText());
     }
 
-    public int getProductPrice() {
-        return getPrice();
+    public static int getProductPriceBeforeDiscount() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:120]: same

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java [hunk 13]
```diff
@@ -1,12 +1,15 @@
...
 import io.qameta.allure.Step;
 
 import static com.codeborne.selenide.Selenide.$x;
 import static com.codeborne.selenide.Condition.visible;
 import static com.codeborne.selenide.Selenide.$;
 
-import static com.softserveinc.ita.pageobjects.ProductModel.*;
+import static com.softserveinc.ita.utils.NumberUtil.parseIntPrice;
 import static java.lang.String.format;
 import static java.time.Duration.ofSeconds;
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:15]: take a look onto getCurrentPriceFromFirstItem if it should be in this page object

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java (hunk 13 above)
Comment (dkrutenko) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:15]: Actually - yes, because we are taking price from first item on products page

Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/ProductPage.java:14]: nope, for my opinion it's search results page
image

**Review by Nikodych — APPROVED** [review]

**Pull request merged**